software/contrib/gates_and_triggers.py

```diff
@@ -40,115 +81,150 @@ def __init__(self):
-            self.on_toggle()
-            ssoled.notify_user_interaction()
-
-    def on_rise(self):
-        """Handle the rising edge of the input signal
-        """
-        self.gate_on = True
-        self.on_incoming_rise_start_time = time.ticks_ms()
-        self.incoming_rise_out.on()
-        self.gate_out.on()
-
-        self.on_toggle()
-
-    def on_fall(self):
-        """Handle the falling edge of the input signal
-        """
-        self.on_incoming_fall_start_time = time.ticks_ms()
-        self.incoming_fall_out.on()
-
-    def on_toggle(self):
-        """Handle toggling the toggle output
-        """
-        self.toggle_on = not self.toggle_on
-        if self.toggle_on:
-            self.toggle_out.on()
-        else:
-            self.toggle_out.off()
-            self.toggle_fall_out.on()
-            self.on_toggle_fall_start_time = time.ticks_ms()
-
-    def tick(self):
-        """Advance the clock and set the outputs high/low as needed
-        """
-        now = time.ticks_ms()
-
-        k1_percent = int(k1.percent() * 100)
-        k2_percent = int(k2.percent() * 100)
-
-        # if the user moved the knobs by more than 1% deactivate the screensaver
-        if abs(self.k1_percent - k1_percent) > 0 or abs(self.k2_percent - k2_percent) > 0:
-            ssoled.notify_user_interaction()
-
-        knob_lvl = k1_percent
-        cv_lvl = ain.percent() * (k2_percent * 0.02 - 1)
-        self.gate_duration = max(self.quadratic_knob(knob_lvl) + cv_lvl * 2000, TRIGGER_DURATION_MS)
-
-        if self.gate_on and time.ticks_diff(now, self.on_incoming_rise_start_time) > self.gate_duration:
-            self.gate_on = False
-            self.gate_out.off()
-            self.gate_fall_out.on()
-            self.on_gate_fall_start_time = now
-
-        if time.ticks_diff(now, self.on_gate_fall_start_time) > TRIGGER_DURATION_MS:
-            self.gate_fall_out.off()
-
-        if time.ticks_diff(now, self.on_incoming_rise_start_time) > TRIGGER_DURATION_MS:
-            self.incoming_rise_out.off()
-
-        if time.ticks_diff(now, self.on_incoming_fall_start_time) > TRIGGER_DURATION_MS:
-            self.incoming_fall_out.off()
-
-        if time.ticks_diff(now, self.on_toggle_fall_start_time) > TRIGGER_DURATION_MS:
-            self.toggle_fall_out.off()
-
-        self.k1_percent = k1_percent
-        self.k2_percent = k2_percent
+            self.force_toggle = True
+            self.last_user_interaction_at = time.ticks_ms()
 
     def quadratic_knob(self, x):
         """Some magic math to give us a quadratic response on the knob percentage
 
-        This gives us 10ms @ 0% and 2000ms @ 100% with greater precision at the higher end
+        This gives us 50ms @ 0% and 1000ms @ 100% with greater precision at the higher end
         where the differences will be more noticeable
+
+        @param x  The value of the knob in the range [0, 100]
         """
+        MIN_GATE = 50
+        MAX_GATE = 1000
```

Comment:
[required] These const definitions would make more sense at a class or script level to make it easier to modify and so they do not get redefined each time the method is called.

Reply:
Done.